apps/bbsync/mixins.py

```diff
@@ -0,0 +1,67 @@
+from django.db import models
+
+from .constants import SYNC_TO_BZ
+from .exceptions import UnsavableModelError
+
+
+class BugzillaSyncMixin(models.Model):
+    """
+    mixin for syncing the model to the Bugzilla
+
+    the sync between Bugzilla and the internal DB is done in a way
+    that the model is first stored to the Bugzilla and then fetched
+    back from there to make sure that the timestaps are kept in sync
+    (otherwise there might be a few seconds differences)
+
+    this mixin does not actually call Model.save at all
+    because it is done when fetched from Bugzilla
+    """
+
+    class Meta:
+        abstract = True
+
+    def save(self, *args, bz_api_key=None, **kwargs):
+        """
+        save the model by storing to Bugzilla and fetching back
+
+        when no Bugzilla API key is provided it is considered to be a call
+        done by a collector or test and thus we perform just regular save
+
+        Bugzilla sync is also conditional based on environment variable
+        """
+        # preliminary save
+        # to link and annotate
+        super().save(*args, **kwargs)
+        self = self.__class__.objects.get(pk=self.pk)
+
+        # check BBSync conditions are met
+        if not SYNC_TO_BZ or bz_api_key is None:
+            return
+
+        # imports here to prevent cycles
+        from collectors.bzimport.collectors import FlawCollector
+        from osidb.models import Affect, Flaw
+
+        from .save import BugzillaSaver
+
+        # Flaw can be directly synced
+        if isinstance(self, Flaw):
+            # sync to Bugzilla
+            bs = BugzillaSaver(self, bz_api_key)
+            self = bs.save()
+            # save in case a new Bugzilla ID was obtained
+            # so the flaw is later matched in BZ import
+            super().save(*args, **kwargs)
+            # fetch from Bugzilla
+            fc = FlawCollector()
+            fc.sync_flaw(self.bz_id)
+
+        # Affect needs to be synced through flaw
+        elif isinstance(self, Affect):
+            self.flaw.save(*args, bz_api_key=bz_api_key, **kwargs)
+
+        else:
+            # no other model can be synced
+            raise UnsavableModelError(
+                f"Instance of {self._meta.model} cannot be synced to Bugzilla"
+            )
```

Comment:
A meta class, mixin, abstract class shouldn't be aware of its implementors, here is what I propose instead:

```python
class BugzillaSyncMixin(...):
    def save(...):
        self.bzsync(...)

    def bzsync(...):
        raise NotImplementedError("Inheritants of BugzillaSyncMixin must implement the bzsync method")


class Flaw(...):
    def bzsync(...):
        # do the sync here
        ...

class Affect(...):
    def bzsync(...):
        self.flaw.save(...)
```

Reply:
Note however that I don't think syncing on flaw save is a good idea, if bz is down or there are network issues we can't save a flaw, even under normal circumstances it would be too slow.

I would rather queue a bbsync job

Reply:
I already tried to explain this in the last comment. This is basically the same what SFM2 does so no degradation. Until we leave Bugzilla we can hardly do it better. The risk of having desynced data between Bugzilla and OSIDB is IMO more serious that two seconds of waiting.

Otherwise your suggestion of unaware abstract class makes sense. I will rework it your way.